android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/SessionPrioritizationMiddleware.kt

```diff
@@ -54,27 +66,59 @@ class SessionPrioritizationMiddleware : Middleware<BrowserState, BrowserAction>
                     updatePriorityIfNeeded(state)
                 }
             }
+            is ContentAction.CheckForFormDataAction -> {
+                val tab = context.state.findTab(action.tabId)
+                if (action.containsFormData) {
+                    tab?.engineState?.engineSession?.updateSessionPriority(HIGH)
+                    logger.info("Update the tab ${tab?.id} priority to ${HIGH.name}")
+                    clearHighPriority(context, tab!!.id)
+
+                } else {
+                    tab?.engineState?.engineSession?.updateSessionPriority(DEFAULT)
+                    logger.info("Update the tab ${tab?.id} priority to ${DEFAULT.name}")
+                }
+            }
+            is ContentAction.ClearHighPrioritySessionAction -> {
+                logger.info("todocathy 3 ClearHighPrioritySessionAction")
+                val tab = context.state.findTab(action.tabId)
+                tab?.engineState?.engineSession?.updateSessionPriority(DEFAULT)
+                logger.info("Update the previous high priority tab ${tab?.id} priority back to ${DEFAULT.name}")
+            }
             else -> {
                 // no-op
             }
         }
     }
 
-    private fun updatePriorityIfNeeded(state: BrowserState) {
+    private fun updatePriorityIfNeeded(state: BrowserState) = mainScope.launch {
         val currentSelectedTab = state.selectedTabId?.let { state.findTab(it) }
         val previousSelectedTab = state.findTab(previousHighestPriorityTabId)
         val currentEngineSession: EngineSession? = currentSelectedTab?.engineState?.engineSession
 
-        // We need to make sure, we alter the previousHighestPriorityTabId, after the session is linked.
+        // We need to make sure we alter the previousHighestPriorityTabId, after the session is linked.
         // So we update the priority on the engine session, as we could get actions where the tab
         // is selected but not linked yet, causing out sync issues,
         // when previousHighestPriorityTabId didn't call updateSessionPriority()
         if (currentEngineSession != null) {
-            previousSelectedTab?.engineState?.engineSession?.updateSessionPriority(DEFAULT)
+            // check for existing form data here and only set DEFAULT if there is
+            previousSelectedTab?.engineState?.engineSession?.checkForFormData()
```

Comment:
I think only this line needs to be wrapped in the `mainScope` call. We can avoid doing excessive work on the UI thread that way (e.g. finding the previous tab) I think, unless there is some reason we are trying to avoid some async situation of finding and reacting to the previous tab?

Reply:
Yes, I should've followed up on the design doc about the change. I had asked you on there if there was a specific action to tell us when a tab is un-selected, because the initial idea I had about using Unlink won't work. Unlink only occurs when a tab is deleted. We need to use the combined action of TabListAction and Link to modify the previous selected tab.

Reply:
This should  achieve it, I was just curios 😅